File: renderers/js/renderer.js
/**
 * Vanilla JS renderer for CoreUI compiler output.
 */
export class CoreUI {
  /**
   * @param {object} output - Parsed JSON output from corec.
   */
  constructor(output) {
    this.output = output || {};
    this.index = new Map();
    this.actionHandlers = new Set();
    this.host = null;
    this.shadowRoot = null;
  }

  /**
   * Renders the CoreUI tree into a shadow root attached to targetElement.
   *
   * @param {HTMLElement} targetElement - Host element for the renderer.
   * @returns {ShadowRoot}
   */
  render(targetElement) {
    if (!(targetElement instanceof HTMLElement)) {
      throw new TypeError("CoreUI.render(targetElement) expects an HTMLElement");
    }

    this.host = targetElement;
    this.shadowRoot =
      targetElement.shadowRoot || targetElement.attachShadow({ mode: "open" });

    this.index.clear();
    this.shadowRoot.replaceChildren();
    this.shadowRoot.appendChild(this.setupTheme());

    if (this.output && this.output.tree) {
      this.shadowRoot.appendChild(this.renderNode(this.output.tree));
    } else {
      this.shadowRoot.appendChild(
        this.createErrorBoundary("Missing CoreUI tree in renderer input.")
      );
    }

    return this.shadowRoot;
  }

  /**
   * Injects theme CSS variables and base styles into the shadow root.
   *
   * @returns {HTMLStyleElement}
   */
  setupTheme() {
    const style = document.createElement("style");
    const theme = this.output && this.output.theme ? this.output.theme : {};
    const declarations = [];

    Object.keys(theme)
      .sort()
      .forEach((key) => {
        const safeKey = this.sanitizeThemeKey(key);
        const resolvedValue = this.resolveThemeDefinitionValue(safeKey, theme[key]);
        if (safeKey && resolvedValue) {
          declarations.push(`--coreui-${safeKey}: ${resolvedValue};`);
        }

        const semanticValue = this.semanticThemeValue(safeKey, theme[key]);
        if (safeKey && semanticValue) {
          declarations.push(`--cui-${safeKey}: ${semanticValue};`);
        }
      });

    style.textContent = `
:host {
  display: block;
  font-family: Arial, Helvetica, sans-serif;
  ${declarations.join("\n  ")}
}
*, *::before, *::after {
  box-sizing: border-box;
}
section, div, span, button, input, img, table, caption, tbody, tr, td, th {
  font: inherit;
}
[data-coreui-error] {
  border: 1px solid #ef4444;
  background: #fef2f2;
  color: #991b1b;
  padding: 0.75rem;
  border-radius: 6px;
  font-family: monospace;
}
`;

    return style;
  }

  /**
   * Renders a single CoreUI node and its children.
   *
   * @param {object} node - CoreUI JSON node.
   * @returns {HTMLElement}
   */
  renderNode(node) {
    if (!node || typeof node !== "object") {
      return this.createErrorBoundary("Invalid node payload.");
    }

    const type = node.type;
    const attrs = this.getAttributes(node);
    let element;

    switch (type) {
      case "View":
        element = document.createElement("section");
        if (attrs.title) {
          const header = document.createElement("header");
          const title = document.createElement("h1");
          title.textContent = attrs.title;
          header.appendChild(title);
          element.appendChild(header);
        }
        break;
      case "Stack":
        element = document.createElement("div");
        break;
      case "Grid":
        element = document.createElement("div");
        break;
      case "Box":
        element = document.createElement("div");
        break;
      case "Text":
        element = document.createElement("span");
        element.textContent = this.asString(attrs.value);
        break;
      case "Input":
        element = this.createInput(node.id, attrs);
        break;
      case "Image":
        element = this.createImage(attrs);
        break;
      case "Trigger":
        element = this.createTrigger(attrs);
        break;
      case "DataTable":
        element = this.createDataTable(attrs);
        break;
      default:
        return this.createErrorBoundary(`Unknown CoreUI component: ${String(type)}`);
    }

    this.decorateElement(element, node);
    this.applyComponentStyles(element, type, attrs);
    if (attrs.hidden === true) {
      element.style.display = "none";
    }

    if (Array.isArray(node.children)) {
      node.children.forEach((child) => {
        element.appendChild(this.renderNode(child));
      });
    }

    return element;
  }

  /**
   * Returns the live DOM element for a CoreUI id, if rendered.
   *
   * @param {string} id - Component id.
   * @returns {HTMLElement|null}
   */
  getById(id) {
    return this.index.get(id) || null;
  }

  /**
   * Registers a callback for Trigger actions.
   *
   * @param {Function} callback - Receives { namespace, call, params, element }.
   * @returns {Function} Unsubscribe function.
   */
  onAction(callback) {
    if (typeof callback !== "function") {
      throw new TypeError("CoreUI.onAction(callback) expects a function");
    }

    this.actionHandlers.add(callback);
    return () => {
      this.actionHandlers.delete(callback);
    };
  }

  createInput(nodeId, attrs) {
    const wrapper = document.createElement("div");
    const inputId = this.asString(nodeId) ? `${this.asString(nodeId)}__input` : "";
    if (attrs.label) {
      const label = document.createElement("label");
      label.textContent = this.asString(attrs.label);
      if (inputId) {
        label.htmlFor = inputId;
      }
      wrapper.appendChild(label);
    }

    const input = document.createElement("input");
    input.type = this.sanitizeHtmlToken(attrs.type) || "text";
    if (inputId) {
      input.id = inputId;
    }
    if (attrs.bind) {
      input.name = this.asString(attrs.bind);
    }
    wrapper.appendChild(input);
    return wrapper;
  }

  createTrigger(attrs) {
    const button = document.createElement("button");
    button.type = "button";
    button.textContent = this.asString(attrs.label);

    const action = attrs.action && typeof attrs.action === "object" ? attrs.action : null;
    if (action) {
      button.addEventListener("click", () => {
        const payload = {
          namespace: this.asString(action.namespace),
          call: this.asString(action.call),
          params: action.params && typeof action.params === "object" ? action.params : {},
          element: button,
        };

        this.actionHandlers.forEach((handler) => {
          handler(payload);
        });
      });
    }

    return button;
  }

  createImage(attrs) {
    const image = document.createElement("img");
    image.src = this.asString(attrs.src);
    image.alt = this.asString(attrs.alt);
    return image;
  }

  createDataTable(attrs) {
    const wrapper = document.createElement("div");
    const table = document.createElement("table");
    if (attrs.source) {
      const caption = document.createElement("caption");
      caption.textContent = this.asString(attrs.source);
      table.appendChild(caption);
    }
    table.appendChild(document.createElement("tbody"));
    wrapper.appendChild(table);
    return wrapper;
  }

  decorateElement(element, node) {
    const id = this.asString(node.id);
    if (id) {
      element.id = id;
      this.index.set(id, element);
    }

    element.dataset.coreuiType = this.asString(node.type);
  }

  applyComponentStyles(element, type, attrs) {
    switch (type) {
      case "Stack":
        element.style.display = "flex";
        element.style.flexDirection = attrs.dir === "h" ? "row" : "column";
        this.applyStyleProperty(element, "gap", attrs.gap);
        this.applyStyleProperty(element, "alignItems", attrs.align);
        break;
      case "Grid":
        element.style.display = "grid";
        this.applyGridTracks(element, "gridTemplateColumns", attrs.cols);
        this.applyGridTracks(element, "gridTemplateRows", attrs.rows);
        this.applyStyleProperty(element, "gap", attrs.gap);
        break;
      case "Box":
        this.applySemanticStyles(element, { interactive: false, elevated: true });
        this.applyVariantStyles(element, attrs.variant);
        this.applyStyleProperty(element, "padding", attrs.padding);
        this.applyStyleProperty(element, "background", attrs.background);
        if (typeof attrs.border === "number") {
          element.style.borderWidth = `${attrs.border}px`;
          element.style.borderStyle = "solid";
        }
        break;
      case "Text":
        this.applyStyleProperty(element, "fontSize", attrs.size);
        this.applyStyleProperty(element, "fontWeight", attrs.weight);
        break;
      case "Image":
        this.applyStyleProperty(element, "width", attrs.width);
        break;
      case "Input":
        this.applySemanticStyles(element.querySelector("input") || element, { interactive: true, elevated: false });
        break;
      case "Trigger":
        this.applySemanticStyles(element, { interactive: true, elevated: true });
        this.applyVariantStyles(element, attrs.variant);
        break;
      default:
        break;
    }

    if (attrs.style) {
      this.applyInlineStyleString(element, attrs.style);
    }
  }

  applyGridTracks(element, property, value) {
    if (!Array.isArray(value) || value.length === 0) {
      return;
    }

    const tracks = value
      .map((item) => this.convertUnit(item, "grid"))
      .filter(Boolean)
      .join(" ");

    if (tracks) {
      element.style[property] = tracks;
    }
  }

  applyInlineStyleString(element, styleText) {
    if (typeof styleText !== "string") {
      return;
    }

    styleText.split(";").forEach((chunk) => {
      const index = chunk.indexOf(":");
      if (index <= 0) {
        return;
      }

      const property = this.sanitizeCssProperty(chunk.slice(0, index));
      const value = chunk.slice(index + 1).trim();
      if (!property) {
        return;
      }

      const camelProperty = property.replace(/-([a-z])/g, (_, letter) =>
        letter.toUpperCase()
      );
      this.applyStyleProperty(element, camelProperty, value, property);
    });
  }

  applyStyleProperty(element, property, value, rawProperty) {
    if (value == null || value === "") {
      return;
    }

    const propertyName =
      rawProperty || property.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`);
    const cssValue = this.resolveStyleValue(propertyName, value);
    if (cssValue) {
      element.style[property] = cssValue;
    }
  }

  resolveStyleValue(property, value) {
    if (this.isColorProperty(property)) {
      const themed = this.resolveThemeToken(value);
      if (themed) {
        return themed;
      }
    }

    if (typeof value === "string") {
      const unitValue = this.convertUnit(value, "literal");
      if (unitValue) {
        return unitValue;
      }
      return this.sanitizeCssValue(value);
    }

    if (typeof value === "number") {
      return String(value);
    }

    return "";
  }

  resolveThemeToken(value) {
    const key = this.sanitizeThemeKey(value);
    if (!key) {
      return "";
    }

    if (this.output && this.output.theme && Object.prototype.hasOwnProperty.call(this.output.theme, key)) {
      return `var(--coreui-${key})`;
    }

    return "";
  }

  resolveThemeDefinitionValue(key, value) {
    const semanticValue = this.semanticThemeValue(key, value);
    if (semanticValue) {
      return semanticValue;
    }

    const tokenKey = this.sanitizeThemeKey(value);
    if (tokenKey && this.themeHas(tokenKey)) {
      return `var(--coreui-${tokenKey})`;
    }

    return this.sanitizeCssValue(value);
  }

  semanticThemeValue(key, value) {
    switch (key) {
      case "radius":
        switch (value) {
          case "none":
            return "0";
          case "sm":
            return "4px";
          case "md":
            return "8px";
          case "lg":
            return "12px";
          case "full":
            return "9999px";
          default:
            return "";
        }
      case "shadow":
        switch (value) {
          case "none":
            return "none";
          case "soft":
            return "0 10px 30px rgba(15, 23, 42, 0.12)";
          case "deep":
            return "0 18px 45px rgba(15, 23, 42, 0.22)";
          default:
            return "";
        }
      case "speed":
        switch (value) {
          case "instant":
            return "all 0s linear";
          case "smooth":
            return "all 180ms ease";
          case "lazy":
            return "all 320ms ease";
          default:
            return "";
        }
      default:
        return "";
    }
  }

  applySemanticStyles(element, options) {
    if (this.themeHas("radius")) {
      element.style.borderRadius = "var(--cui-radius)";
    }
    if (options && options.elevated && this.themeHas("shadow")) {
      element.style.boxShadow = "var(--cui-shadow)";
    }
    if (options && options.interactive && this.themeHas("speed")) {
      element.style.transition = "var(--cui-speed)";
    }
  }

  applyVariantStyles(element, variant) {
    const safeVariant = this.sanitizeThemeKey(variant);
    const primary = this.resolveThemeToken("primary");
    const text = this.resolveThemeToken("text") || "inherit";

    if (!safeVariant || !primary) {
      return;
    }

    switch (safeVariant) {
      case "primary":
        element.style.background = primary;
        element.style.borderWidth = "1px";
        element.style.borderStyle = "solid";
        element.style.borderColor = primary;
        element.style.color = text;
        break;
      case "secondary":
      case "outline":
        element.style.background = "transparent";
        element.style.borderWidth = "1px";
        element.style.borderStyle = "solid";
        element.style.borderColor = primary;
        element.style.color = primary;
        break;
      case "ghost":
        element.style.background = "transparent";
        element.style.borderWidth = "1px";
        element.style.borderStyle = "solid";
        element.style.borderColor = "transparent";
        element.style.color = primary;
        break;
      default:
        break;
    }
  }

  convertUnit(value, context) {
    if (typeof value !== "string") {
      return "";
    }

    const trimmed = value.trim();
    if (!trimmed) {
      return "";
    }

    if (trimmed === "auto" || trimmed.endsWith("px") || trimmed.endsWith("%")) {
      return this.sanitizeCssValue(trimmed);
    }

    if (trimmed.endsWith("*")) {
      const weight = trimmed.slice(0, -1).trim() || "1";
      if (!/^\d+(\.\d+)?$/.test(weight)) {
        return "";
      }
      if (context === "grid") {
        return `${weight}fr`;
      }
      return weight;
    }

    return "";
  }

  getAttributes(node) {
    if (!node || typeof node !== "object") {
      return {};
    }
    if (node.attributes && typeof node.attributes === "object") {
      return node.attributes;
    }
    if (node.props && typeof node.props === "object") {
      return node.props;
    }
    return {};
  }

  createErrorBoundary(message) {
    const error = document.createElement("div");
    error.dataset.coreuiError = "true";
    error.textContent = message;
    return error;
  }

  sanitizeCssProperty(value) {
    if (typeof value !== "string") {
      return "";
    }

    const trimmed = value.trim().toLowerCase();
    return /^[a-z-]+$/.test(trimmed) ? trimmed : "";
  }

  sanitizeCssValue(value) {
    if (typeof value !== "string") {
      return "";
    }

    const trimmed = value.trim();
    if (!trimmed) {
      return "";
    }

    if (/["';{}<>[\]\\`\n\r]/.test(trimmed)) {
      return "";
    }

    return /^[a-zA-Z0-9#%.(),\-_/+\s]+$/.test(trimmed) ? trimmed : "";
  }

  sanitizeThemeKey(value) {
    if (typeof value !== "string") {
      return "";
    }
    const trimmed = value.trim();
    return /^[a-zA-Z0-9_-]+$/.test(trimmed) ? trimmed : "";
  }

  sanitizeHtmlToken(value) {
    if (typeof value !== "string") {
      return "";
    }
    const trimmed = value.trim();
    return /^[a-zA-Z0-9:_-]+$/.test(trimmed) ? trimmed : "";
  }

  isColorProperty(property) {
    return (
      property === "background" ||
      property === "background-color" ||
      property === "color" ||
      property === "border-color" ||
      property === "fill" ||
      property === "stroke"
    );
  }

  themeHas(key) {
    return (
      this.output &&
      this.output.theme &&
      Object.prototype.hasOwnProperty.call(this.output.theme, key)
    );
  }

  asString(value) {
    return value == null ? "" : String(value);
  }
}

export default CoreUI;

File: renderers/goth/renderer.go
package goth

import (
	"strings"

	"github.com/a-h/templ"
	"github.com/valueforvalue/coreui/pkg/ast"
	"github.com/valueforvalue/coreui/pkg/registry"
)

// Render converts a CoreUI AST node into a templ component tree.
//
// To serve the result with net/http, call Render on the root node and render
// the returned component inside your HTTP handler or page template.
func Render(node ast.Node) templ.Component {
	return RenderWithTheme(node, nil)
}

// RenderWithTheme converts a CoreUI AST node into a templ component tree and
// injects CSS custom properties for the provided theme tokens.
func RenderWithTheme(node ast.Node, theme map[string]string) templ.Component {
	return renderNode(&node, theme)
}

func renderNode(node *ast.Node, theme map[string]string) templ.Component {
	if node == nil {
		return templ.Raw("")
	}

	children := renderChildren(node.Children, theme)
	id := node.ID()

	switch node.Type {
	case "View":
		return viewComponent(id, stringAttribute(node, "title"), baseStyle(node, theme), RenderTheme(theme), children)
	case "Stack":
		direction := "column"
		if stringAttribute(node, "dir") == "h" {
			direction = "row"
		}

		decls := []styleDecl{
			styleDeclFor("display", "flex"),
			styleDeclFor("flex-direction", direction),
		}
		if gap := UnitToCSS(unitAttribute(node, "gap"), UnitContextLiteral); gap != "" {
			decls = append(decls, styleDeclFor("gap", gap))
		}
		if align := sanitizeCSSToken(stringAttribute(node, "align")); align != "" {
			decls = append(decls, styleDeclFor("align-items", align))
		}

		return stackComponent(id, baseStyle(node, theme, decls...), children)
	case "Grid":
		decls := []styleDecl{
			styleDeclFor("display", "grid"),
		}
		if cols := unitArrayAttribute(node, "cols", UnitContextGridTrack); cols != "" {
			decls = append(decls, styleDeclFor("grid-template-columns", cols))
		}
		if rows := unitArrayAttribute(node, "rows", UnitContextGridTrack); rows != "" {
			decls = append(decls, styleDeclFor("grid-template-rows", rows))
		}
		if gap := UnitToCSS(unitAttribute(node, "gap"), UnitContextLiteral); gap != "" {
			decls = append(decls, styleDeclFor("gap", gap))
		}

		return gridComponent(id, baseStyle(node, theme, decls...), children)
	case "Box":
		decls := make([]styleDecl, 0, 8)
		decls = append(decls, semanticStyleDecls(theme, false, true)...)
		decls = append(decls, variantStyleDecls(stringAttribute(node, "variant"), theme)...)
		if padding := UnitToCSS(unitAttribute(node, "padding"), UnitContextLiteral); padding != "" {
			decls = append(decls, styleDeclFor("padding", padding))
		}
		if background := stringAttribute(node, "background"); background != "" {
			decls = append(decls, styleDeclFor("background", background))
		}
		if border, ok := intAttribute(node, "border"); ok && border > 0 {
			decls = append(decls,
				styleDeclFor("border-width", formatInt(border)+"px"),
				styleDeclFor("border-style", "solid"),
			)
		}

		return boxComponent(id, baseStyle(node, theme, decls...), children)
	case "Text":
		decls := make([]styleDecl, 0, 2)
		if size := UnitToCSS(unitAttribute(node, "size"), UnitContextLiteral); size != "" {
			decls = append(decls, styleDeclFor("font-size", size))
		}
		if weight := sanitizeCSSToken(stringAttribute(node, "weight")); weight != "" {
			decls = append(decls, styleDeclFor("font-weight", weight))
		}

		return textComponent(id, stringAttribute(node, "value"), baseStyle(node, theme, decls...))
	case "Input":
		inputType := sanitizeHTMLToken(stringAttribute(node, "type"))
		if inputType == "" {
			inputType = "text"
		}
		decls := semanticStyleDecls(theme, true, false)
		return inputComponent(
			id,
			stringAttribute(node, "label"),
			stringAttribute(node, "bind"),
			inputType,
			baseStyle(node, theme, decls...),
		)
	case "Image":
		decls := make([]styleDecl, 0, 1)
		if width := UnitToCSS(unitAttribute(node, "width"), UnitContextLiteral); width != "" {
			decls = append(decls, styleDeclFor("width", width))
		}
		return imageComponent(
			id,
			stringAttribute(node, "src"),
			stringAttribute(node, "alt"),
			baseStyle(node, theme, decls...),
		)
	case "Trigger":
		action, _ := actionAttribute(node, "action")
		payload := encodeActionRequest(actionRequestFromAction(action))
		decls := make([]styleDecl, 0, 8)
		decls = append(decls, semanticStyleDecls(theme, true, true)...)
		decls = append(decls, variantStyleDecls(stringAttribute(node, "variant"), theme)...)
		return triggerComponent(
			id,
			stringAttribute(node, "label"),
			sanitizeHTMLToken(stringAttribute(node, "variant")),
			payload,
			baseStyle(node, theme, decls...),
		)
	case "DataTable":
		return dataTableComponent(
			id,
			stringAttribute(node, "source"),
			boolAttribute(node, "selectable"),
			baseStyle(node, theme),
			children,
		)
	case "Graph":
		values, reference := graphDataAttribute(node, "data")
		decls := []styleDecl{
			styleDeclFor("display", "flex"),
			styleDeclFor("flex-direction", "column"),
			styleDeclFor("gap", "12px"),
			styleDeclFor("width", "100%"),
		}
		height := UnitToCSS(unitAttribute(node, "height"), UnitContextLiteral)
		if height == "" {
			height = "240px"
		}
		return graphComponent(
			id,
			baseStyle(node, theme, decls...),
			stringAttribute(node, "type"),
			values,
			stringArrayAttribute(node, "labels"),
			reference,
			graphColorValue(stringAttribute(node, "color"), theme),
			graphRadiusValue(theme),
			height,
		)
	default:
		if registry.IsPluginComponent(node.Type) {
			return pluginComponent(id, node.Type, baseStyle(node, theme), children)
		}
		return unknownComponent(id, node.Type, baseStyle(node, theme), children)
	}
}

func renderChildren(children []*ast.Node, theme map[string]string) []templ.Component {
	if len(children) == 0 {
		return nil
	}

	components := make([]templ.Component, 0, len(children))
	for _, child := range children {
		components = append(components, renderNode(child, theme))
	}
	return components
}

func stringAttribute(node *ast.Node, key string) string {
	if node == nil {
		return ""
	}
	value, ok := node.Attributes[key]
	if !ok || value.Kind != ast.StringKind {
		return ""
	}
	text, _ := value.Data.(string)
	return text
}

func unitAttribute(node *ast.Node, key string) string {
	if node == nil {
		return ""
	}
	value, ok := node.Attributes[key]
	if !ok || value.Kind != ast.UnitKind {
		return ""
	}
	text, _ := value.Data.(string)
	return text
}

func unitArrayAttribute(node *ast.Node, key string, context UnitContext) string {
	if node == nil {
		return ""
	}
	value, ok := node.Attributes[key]
	if !ok || value.Kind != ast.ArrayKind {
		return ""
	}
	items, _ := value.Data.([]ast.Value)
	if len(items) == 0 {
		return ""
	}

	parts := make([]string, 0, len(items))
	for _, item := range items {
		if item.Kind != ast.UnitKind {
			continue
		}
		text, _ := item.Data.(string)
		if css := UnitToCSS(text, context); css != "" {
			parts = append(parts, css)
		}
	}

	return strings.Join(parts, " ")
}

func boolAttribute(node *ast.Node, key string) bool {
	if node == nil {
		return false
	}
	value, ok := node.Attributes[key]
	if !ok || value.Kind != ast.BoolKind {
		return false
	}
	flag, _ := value.Data.(bool)
	return flag
}

func intAttribute(node *ast.Node, key string) (int64, bool) {
	if node == nil {
		return 0, false
	}
	value, ok := node.Attributes[key]
	if !ok || value.Kind != ast.IntKind {
		return 0, false
	}
	number, ok := value.Data.(int64)
	return number, ok
}

func formatInt(value int64) string {
	if value == 0 {
		return "0"
	}

	buf := [20]byte{}
	pos := len(buf)
	for value > 0 {
		pos--
		buf[pos] = byte('0' + (value % 10))
		value /= 10
	}

	return string(buf[pos:])
}

func actionAttribute(node *ast.Node, key string) (ast.Action, bool) {
	if node == nil {
		return ast.Action{}, false
	}
	value, ok := node.Attributes[key]
	if !ok || value.Kind != ast.ActionKind {
		return ast.Action{}, false
	}
	action, ok := value.Data.(ast.Action)
	return action, ok
}

func stringArrayAttribute(node *ast.Node, key string) []string {
	if node == nil {
		return nil
	}
	value, ok := node.Attributes[key]
	if !ok || value.Kind != ast.ArrayKind {
		return nil
	}
	items, _ := value.Data.([]ast.Value)
	if len(items) == 0 {
		return nil
	}

	out := make([]string, 0, len(items))
	for _, item := range items {
		if item.Kind != ast.StringKind {
			continue
		}
		text, _ := item.Data.(string)
		out = append(out, text)
	}
	return out
}

func graphDataAttribute(node *ast.Node, key string) ([]float64, string) {
	if node == nil {
		return nil, ""
	}
	value, ok := node.Attributes[key]
	if !ok {
		return nil, ""
	}

	switch value.Kind {
	case ast.StringKind:
		text, _ := value.Data.(string)
		return nil, text
	case ast.ArrayKind:
		items, _ := value.Data.([]ast.Value)
		out := make([]float64, 0, len(items))
		for _, item := range items {
			switch item.Kind {
			case ast.IntKind:
				number, _ := item.Data.(int64)
				out = append(out, float64(number))
			case ast.NumberKind:
				number, _ := item.Data.(float64)
				out = append(out, number)
			}
		}
		return out, ""
	default:
		return nil, ""
	}
}

func graphColorValue(value string, theme map[string]string) string {
	if token := resolveThemeToken(value, theme); token != "" {
		return token
	}
	if token := resolveThemeToken("primary", theme); token != "" {
		return token
	}
	if sanitized := sanitizeCSSValue(value); sanitized != "" {
		return sanitized
	}
	return "#6366f1"
}

func graphRadiusValue(theme map[string]string) string {
	if radius := semanticTokenCSSValue("radius", theme["radius"]); radius != "" {
		return radius
	}
	return "8px"
}
